compiler/src/iree/compiler/Codegen/Common/IREEExpandStridedMetadata.cpp

```diff
@@ -0,0 +1,265 @@
+// Copyright 2023 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#define DEBUG_TYPE "iree-codegen-expand-strided-metadata"
+
+#include "iree/compiler/Codegen/Dialect/IREECodegenDialect.h"
+#include "iree/compiler/Codegen/Dialect/UKernelOps.h"
+#include "iree/compiler/Codegen/PassDetail.h"
+#include "iree/compiler/Codegen/Passes.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/MemRef/Transforms/Transforms.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir {
+namespace iree_compiler {
+
+namespace {
+/// Helper struct to return the offset, sizes and strides
+/// of a `source` of a `memref.extract_strided_metadata` op.
+struct DescriptorInfo {
+  OpFoldResult offset;
+  SmallVector<OpFoldResult> sizes;
+  SmallVector<OpFoldResult> strides;
+};
+}  // namespace
+
+/// Returns an AffineMap for an add or a mul.
+static AffineMap getAddMap(MLIRContext *context) {
+  AffineExpr s0, s1;
+  bindSymbols(context, s0, s1);
+  return AffineMap::get(0, 2, s0 + s1);
+}
+static AffineMap getMulMap(MLIRContext *context) {
+  AffineExpr s0, s1;
+  bindSymbols(context, s0, s1);
+  return AffineMap::get(0, 2, s0 * s1);
+}
+
+/// Returns the strides based on the sizes assuming that the `memref`
+/// has default layout, i.e. it is not a result of a subview.
+static SmallVector<OpFoldResult> getStridesFromSizes(
+    RewriterBase &rewriter, Location loc, ArrayRef<OpFoldResult> sizes) {
+  if (sizes.size() == 0) {
+    return {};
+  }
+  SmallVector<OpFoldResult> strides(sizes.size());
+  strides.back() = rewriter.getIndexAttr(1);
+  if (sizes.size() == 1) {
+    return strides;
+  }
+  AffineMap mulMap = getMulMap(rewriter.getContext());
+  for (int i = sizes.size() - 2; i >= 0; --i) {
+    strides[i] = makeComposedFoldedAffineApply(rewriter, loc, mulMap,
+                                               {strides[i + 1], sizes[i + 1]});
+  }
+  return strides;
+}
+
+static FailureOr<DescriptorInfo> resolveBufferDescriptorForInterfaceBinding(
```

Comment:
This function won't fail? Don't need to have the wrapping `FailureOr` to me.

Reply:
I'd prefer to keep it that way... It doesnt fail today, but just keeping it this way to make sure any failures do get propagated.